README.md
# Jitsi Magnify

An authentication, room and meeting management system for Jitsi based on Django/React.

Jitsi Magnify is built with [ReactJS](https://fr.reactjs.org/) for the frontend, and
[Django](https://www.djangoproject.com/) for the backend.

## Getting started

### Prerequisite

Make sure you have a recent version of Docker and
[Docker Compose](https://docs.docker.com/compose/install) installed on your laptop:

```bash
$ docker -v
  Docker version 20.10.2, build 2291f61

$ docker compose version
  Docker Compose version v2.17.3 
```

>⚠️ You may need to run the following commands with `sudo` but this can be
>avoided by assigning your user to the `docker` group.

### Project bootstrap

The easiest way to start working on the project is to use our `Makefile` :
```bash
$ make bootstrap
```

This command builds the `app` container, installs dependencies and performs database migrations.
It's a good idea to use this command each time you are pulling code from the project repository
to avoid dependency-releated or migration-releated issues.

When the command stops, check that all services are running as expected:

```bash
$ docker compose ps
```

You should now be able to access the demo site at [localhost:8070](http://localhost:8070).

Finally, you can see all available commands in our `Makefile` with :

```bash
$ make help
```

### Django admin

You can access the Django admin site at [localhost:8071/admin](http://localhost:8071/admin/).

To access the Django admin, you will first need to create a superuser account:

```bash
$ make superuser
```

## Running Magnify in production

### Configure a Jitsi instance

Before running Magnify, you will need a Jitsi instance with JWT authentication activated:

```
ENABLE_AUTH=1
AUTH_TYPE=jwt
JWT_APP_ID=magnify
JWT_APP_SECRET={JWT_JITSI_APP_SECRET}
```

In the Prosody configuration, you should also set the variable:
`XMPP_DOMAIN={JWT_JITSI_XMPP_DOMAIN}`.

The `JWT_JITSI_APP_SECRET` and `JWT_JITSI_XMPP_DOMAIN` variables should be set to the same value
in your Jitsi instance and in Magnify.

### Configure Magnify

The easiest way to run Magnify in production is to use the [official Docker image][1].

Configuration is done via environment variables as detailed in our
[configuration guide](docs/env.md).


## Frontend

#### Architecture

The front project is split into two parts.

- A first part in `src/frontend/packages/core` contains all the components, services, repositories, and even complete
  pages required to build a magnify application. It also includes an AppRouter component that creates an app and
  its default routes


- Then a sandbox application (`src/frontend/sandbox`) which aims to demonstrate how to use the core package.

#### Configuration variables

A set of configuration variables is required. All variables can be configured directly through the environment
variables of the Django project. They are served to the client via the `/config.json` API route.

Here is the list of all the available variables :

```
{
  "API_URL": "http://localhost:8071/api" // URL of magnify api,
  "JITSI_DOMAIN": "exemple.test",
  "KEYCLOAK_CLIENT_ID": "magnify-front",
  "KEYCLOAK_EXPIRATION_SECONDS": 1800,
  "KEYCLOAK_REALM": "magnify",
  "MAGNIFY_SHOW_REGISTER_LINK": true,
  "KEYCLOAK_URL": "http://localhost:8080"
}

```
You can mock these variables by adding a `config.json` file in the public folder of the sandbox application.

#### Development mode

We have added a compilation option that allows the compiler to directly access the project sources when it encounters
an import from the `@openfun/jitsi-magnify` package.

As a result, to use package components in the sandbox, you don't need to build the package. You just need to export them.

To learn how to export new components, please open the `src/frontend/packages/core/index.ts` file.

You can now navigate to the `src/frontend/sandbox` folder and run the `yarn dev` command directly. Hot reload will
work when you modify a component in the `package/core`.


#### Customization

In order to make magnify customizable, we opted to add the @openfun/cunningham-react package.
[cunningham documentation](https://github.com/openfun/cunningham)

However, cunningham does not contain all the necessary components. So we are still using Grommet for now. We need to do
a mapping between the different cunningham tokens and the Grommet theme configuration.

To see how this mapping works, go to `src/frontend/packages/core/themes/theme.ts` file.


## Contributing

This project is intended to be community-driven, so please, do not hesitate to
get in touch if you have any question related to our implementation or design
decisions.

We try to raise our code quality standards and expect contributors to follow
the recommandations from our
[handbook](https://handbook.openfun.fr).

## License

This work is released under the MIT License (see [LICENSE](./LICENSE)).

[1]: https://hub.docker.com/r/fundocker/jitsi-magnify


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [Unrealeased]

### Added

- Improve a few french translations issues in the fr-Fr.json
- Switch the first letter of the help comments in the makefile to uppercase
- Upgrade docker compose to latest version
- Enable CORS Headers
- Add Multilingual
- Add default language management with LANGUAGE_CODE
- Use the backend's jitsi name to match the JWT token and set the subject to
  the actual room name.
- Fix the css of scrolling of the `SimpleLayout` component
- Use a "back" link instead of a breadcrumb in the room settings page
- Add default language management with LANGUAGE_CODE
- Use the backend's jitsi name to match the JWT token and set the subject to
  the actual room name.
- Allow hiding the register link
- Allow to customize frontend application through Cunningham tokens


sandbox/settings.py
"""
Django settings for magnify project.
"""
import json
import os

from django.utils.translation import gettext_lazy as _

# pylint: disable=ungrouped-imports
import sentry_sdk
from configurations import Configuration, values
from sentry_sdk.integrations.django import DjangoIntegration

from magnify.apps.core.settings.mixins import MagnifyCoreConfigurationMixin

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join("/", "data")


def get_release():
    """Get the current release of the application.

    By release, we mean the release from the version.json file à la Mozilla [1]
    (if any). If this file has not been found, it defaults to "NA".

    [1]
    https://github.com/mozilla-services/Dockerflow/blob/master/docs/version_object.md
    """
    # Try to get the current release from the version.json file generated by the
    # CI during the Docker image build
    try:
        with open(os.path.join(BASE_DIR, "version.json"), encoding="utf8") as version:
            return json.load(version)["version"]
    except FileNotFoundError:
        return "NA"  # Default: not available


class Base(MagnifyCoreConfigurationMixin, Configuration):
    """
    This is the base configuration every configuration (aka environnement) should inherit from. It
    is recommended to configure third-party applications by creating a configuration mixins in
    ./configurations and compose the Base configuration with those mixins.

    It depends on an environment variable that SHOULD be defined:

    * DJANGO_SECRET_KEY

    You may also want to override default configuration by setting the following environment
    variables:

    * DJANGO_SENTRY_DSN
    * magnify_ES_HOST
    * DB_NAME
    * DB_HOST
    * DB_PASSWORD
    * DB_USER
    """

    DEBUG = False

    SITE_ID = 1

    # Security
    ALLOWED_HOSTS = []
    CSRF_TRUSTED_ORIGINS = values.ListValue([])
    SECRET_KEY = values.Value(None)

    # CORS headers
    CORS_ALLOWED_ORIGINS = values.ListValue([])

    # System check reference:
    # https://docs.djangoproject.com/en/2.2/ref/checks/#security
    SILENCED_SYSTEM_CHECKS = values.ListValue(
        [
            # Allow the X_FRAME_OPTIONS to be set to "SAMEORIGIN"
            "security.W019"
        ]
    )

    REST_FRAMEWORK = {
        "ALLOWED_VERSIONS": ("1.0",),
        "DEFAULT_AUTHENTICATION_CLASSES": values.ListValue(
            ["rest_framework_simplejwt.authentication.JWTAuthentication"],
            environ_name="DRF_DEFAULT_AUTHENTICATION_CLASSES",
            environ_prefix=None,
        ),
        "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.PageNumberPagination",
        "PAGE_SIZE": 100,
        "DEFAULT_VERSION": "1.0",
        "DEFAULT_VERSIONING_CLASS": "rest_framework.versioning.URLPathVersioning",
        "EXCEPTION_HANDLER": "magnify.apps.core.api.exception_handler",
    }

    # Frontend
    FRONTEND_CONFIGURATION = {
        "API_URL": values.Value(environ_name="MAGNIFY_API_URL", environ_prefix=None),
        "SHOW_REGISTER_LINK": values.BooleanValue(
            True, environ_name="MAGNIFY_SHOW_REGISTER_LINK", environ_prefix=None
        ),
        "KEYCLOAK_URL": values.Value(environ_name="KEYCLOAK_URL", environ_prefix=None),
        "KEYCLOAK_REALM": values.Value(
            "magnify", environ_name="KEYCLOAK_REALM", environ_prefix=None
        ),
        "KEYCLOAK_CLIENT_ID": values.Value(
            "magnify-front", environ_name="KEYCLOAK_CLIENT_ID", environ_prefix=None
        ),
        "KEYCLOAK_EXPIRATION_SECONDS": values.IntegerValue(
            30 * 60,
            environ_name="KEYCLOAK_EXPIRATION_SECONDS",
            environ_prefix=None,
        ),
    }

    # Application definition
    ROOT_URLCONF = "urls"
    WSGI_APPLICATION = "wsgi.application"

    AUTH_USER_MODEL = "core.User"

    JITSI_CONFIGURATION = {
        "jitsi_domain": values.Value(environ_name="JITSI_DOMAIN", environ_prefix=None),
        "jitsi_app_id": values.Value(environ_name="JITSI_APP_ID", environ_prefix=None),
        "jitsi_secret_key": values.Value(
            environ_name="JITSI_SECRET_KEY", environ_prefix=None
        ),
        "jitsi_xmpp_domain": values.Value(
            environ_name="JITSI_XMPP_DOMAIN", environ_prefix=None
        ),
        "jitsi_guest_avatar": values.Value(
            "", environ_name="JITSI_GUEST_AVATAR", environ_prefix=None
        ),
        "jitsi_guest_username": values.Value(
            "Guest", environ_name="JITSI_GUEST_USERNAME", environ_prefix=None
        ),
        "jitsi_token_expiration_seconds": values.Value(
            300, environ_name="JITSI_TOKEN_EXPIRATION_SECONDS", environ_prefix=None
        ),
    }

    JITSI_ROOM_PREFIX = values.Value(
        "", environ_name="MAGNIFY_JITSI_ROOM_PREFIX", environ_prefix=None
    )
    DEFAULT_ROOM_IS_PUBLIC = values.BooleanValue(
        True, environ_name="MAGNIFY_DEFAULT_ROOM_IS_PUBLIC", environ_prefix=None
    )
    ALLOW_UNREGISTERED_ROOMS = values.BooleanValue(
        True, environ_name="MAGNIFY_ALLOW_UNREGISTERED_ROOMS", environ_prefix=None
    )
    ALLOW_API_USER_CREATE = values.BooleanValue(
        False, environ_name="MAGNIFY_ALLOW_API_USER_CREATE", environ_prefix=None
    )

    # Database
    DATABASES = {
        "default": {
            "ENGINE": values.Value(
                "django.db.backends.postgresql_psycopg2",
                environ_name="DB_ENGINE",
                environ_prefix=None,
            ),
            "NAME": values.Value(
                "magnify", environ_name="DB_NAME", environ_prefix=None
            ),
            "USER": values.Value(
                "magnify", environ_name="DB_USER", environ_prefix=None
            ),
            "PASSWORD": values.Value(environ_name="DB_PASSWORD", environ_prefix=None),
            "HOST": values.Value(
                "postgresql", environ_name="DB_HOST", environ_prefix=None
            ),
            "PORT": values.Value(5432, environ_name="DB_PORT", environ_prefix=None),
        }
    }
    DEFAULT_AUTO_FIELD = "django.db.models.AutoField"
    MIGRATION_MODULES = {}

    # Static files (CSS, JavaScript, Images)
    STATIC_URL = "/static/"
    MEDIA_URL = "/media/"
    MEDIA_ROOT = os.path.join(DATA_DIR, "media")
    STATIC_ROOT = os.path.join(DATA_DIR, "static")

    # Simple JWT
    SIMPLE_JWT = {
        "ALGORITHM": values.Value(
            "RS256", environ_name="MAGNIFY_JWT_ALGORITHM", environ_prefix=None
        ),
        "JWK_URL": values.Value(
            None, environ_name="MAGNIFY_JWT_JWK_URL", environ_prefix=None
        ),
        "SIGNING_KEY": values.Value(
            None, environ_name="MAGNIFY_JWT_SIGNING_KEY", environ_prefix=None
        ),
        "VERIFYING_KEY": values.Value(
            None, environ_name="MAGNIFY_JWT_VERIFYING_KEY", environ_prefix=None
        ),
        "AUTH_HEADER_TYPES": ("Bearer",),
        "AUTH_HEADER_NAME": "HTTP_AUTHORIZATION",
        "TOKEN_TYPE_CLAIM": "typ",
        "USER_ID_FIELD": "jwt_sub",
        "USER_ID_CLAIM": "sub",
        "AUTH_TOKEN_CLASSES": ("magnify.apps.core.tokens.BearerToken",),
    }
    JWT_USER_FIELDS_SYNC = values.DictValue(
        {
            "email": "email",
            "name": "name",
            "username": "preferred_username",
        },
        environ_name="MAGNIFY_JWT_USER_FIELDS_SYNC",
        environ_prefix=None,
    )
    JWT_USER_DEVICE_AUDIENCES = values.ListValue(
        [],
        environ_name="MAGNIFY_JWT_USER_DEVICE_AUDIENCES",
        environ_prefix=None,
    )

    # Login/registration related settings
    LOGIN_REDIRECT_URL = "/"
    LOGOUT_REDIRECT_URL = "/"

    AUTHENTICATION_BACKENDS = ("django.contrib.auth.backends.ModelBackend",)

    # Internationalization
    TIME_ZONE = "UTC"
    USE_I18N = True
    USE_L10N = True
    USE_TZ = True

    # Templates
    TEMPLATES = [
        {
            "BACKEND": "django.template.backends.django.DjangoTemplates",
            "DIRS": [os.path.join(BASE_DIR, "templates")],
            "OPTIONS": {
                "context_processors": [
                    "django.contrib.auth.context_processors.auth",
                    "django.contrib.messages.context_processors.messages",
                    "django.template.context_processors.i18n",
                    "django.template.context_processors.debug",
                    "django.template.context_processors.request",
                    "django.template.context_processors.media",
                    "django.template.context_processors.csrf",
                    "django.template.context_processors.tz",
                    "django.template.context_processors.static",
                ],
                "loaders": [
                    "django.template.loaders.filesystem.Loader",
                    "django.template.loaders.app_directories.Loader",
                ],
            },
        }
    ]

    MIDDLEWARE = (
        "django.middleware.security.SecurityMiddleware",
        "whitenoise.middleware.WhiteNoiseMiddleware",
        "django.contrib.sessions.middleware.SessionMiddleware",
        "django.middleware.csrf.CsrfViewMiddleware",
        "django.contrib.auth.middleware.AuthenticationMiddleware",
        "django.contrib.messages.middleware.MessageMiddleware",
        "django.middleware.locale.LocaleMiddleware",
        "django.middleware.common.CommonMiddleware",
        "django.middleware.clickjacking.XFrameOptionsMiddleware",
        "corsheaders.middleware.CorsMiddleware",
        "dockerflow.django.middleware.DockerflowMiddleware",
    )

    # Swagger
    SWAGGER_SETTINGS = {
        "SECURITY_DEFINITIONS": {
            "Bearer": {"type": "apiKey", "name": "Authorization", "in": "header"},
        }
    }

    # Django applications from the highest priority to the lowest
    INSTALLED_APPS = (
        # magnify stuff
        "magnify.apps.core",
        "magnify",
        # Third party apps
        "corsheaders",
        "dockerflow.django",
        "parler",
        "rest_framework",
        "drf_yasg",
        # Django
        "django.contrib.auth",
        "django.contrib.contenttypes",
        "django.contrib.sessions",
        "django.contrib.admin",
        "django.contrib.sites",
        "django.contrib.sitemaps",
        "django.contrib.staticfiles",
        "django.contrib.messages",
        "django.contrib.humanize",
    )

    # Languages
    # - Django
    LANGUAGE_CODE = values.Value("en")

    # Careful! Languages should be ordered by priority, as this tuple is used to get
    # fallback/default languages throughout the app.
    # Use "en" as default as it is the language that is most likely to be spoken by any visitor
    # when their preferred language, whatever it is, is unavailable
    LANGUAGES = (("en", _("English")), ("fr", _("French")))

    # Logging
    LOGGING = {
        "version": 1,
        "disable_existing_loggers": True,
        "formatters": {
            "verbose": {
                "format": "%(levelname)s %(asctime)s %(module)s "
                "%(process)d %(thread)d %(message)s"
            }
        },
        "handlers": {
            "console": {
                "level": "DEBUG",
                "class": "logging.StreamHandler",
                "formatter": "verbose",
            }
        },
        "loggers": {
            "django.db.backends": {
                "level": "ERROR",
                "handlers": ["console"],
                "propagate": False,
            }
        },
    }

    # Cache
    CACHES = {
        "default": {
            "BACKEND": values.Value(
                "django.core.cache.backends.locmem.LocMemCache",
                environ_name="CACHE_DEFAULT_BACKEND",
                environ_prefix=None,
            ),
            "LOCATION": values.Value(
                "", environ_name="CACHE_DEFAULT_LOCATION", environ_prefix=None
            ),
            "OPTIONS": values.DictValue(
                {}, environ_name="CACHE_DEFAULT_OPTIONS", environ_prefix=None
            ),
        },
    }

    # Sentry
    SENTRY_DSN = values.Value(None, environ_name="SENTRY_DSN")

    @classmethod
    def _get_environment(cls):
        """Environment in which the application is launched."""
        return cls.__name__.lower()

    # pylint: disable=invalid-name
    @property
    def ENVIRONMENT(self):
        """Environment in which the application is launched."""
        return self._get_environment()

    # pylint: disable=invalid-name
    @property
    def RELEASE(self):
        """
        Return the release information.

        Delegate to the module function to enable easier testing.
        """
        return get_release()

    @classmethod
    def post_setup(cls):
        """Post setup configuration.
        This is the place where you can configure settings that require other
        settings to be loaded.
        """
        super().post_setup()

        # The SENTRY_DSN setting should be available to activate sentry for an environment
        if cls.SENTRY_DSN is not None:
            sentry_sdk.init(  # pylint: disable=abstract-class-instantiated
                dsn=cls.SENTRY_DSN,
                environment=cls._get_environment(),
                release=get_release(),
                integrations=[DjangoIntegration()],
            )
            with sentry_sdk.configure_scope() as scope:
                scope.set_extra("application", "backend")


class Build(Base):
    """Build environment settings"""

    SECRET_KEY = "ThisIsAnExampleKeyForBuildPurposeOnly"  # nosec
    JWT_JITSI_SECRET_KEY = "ThisIsAnExampleKeyForBuildPurposeOnly"  # nosec

    STORAGES = {
        "staticfiles": {
            "BACKEND": str(
                values.Value("whitenoise.storage.CompressedManifestStaticFilesStorage")
            ),
        },
    }


class Development(Base):
    """
    Development environment settings

    We set DEBUG to True and configure the server to respond from all hosts.
    """

    DEBUG = True
    ALLOWED_HOSTS = ["*"]
    CORS_ALLOW_ALL_ORIGINS = True
    CSRF_TRUSTED_ORIGINS = ["http://localhost:8071"]

    @classmethod
    def post_setup(cls):
        """Post setup configuration.
        Activate local Keycloak as authentication backend for development.
        """
        super().post_setup()
        cls.REST_FRAMEWORK["DEFAULT_AUTHENTICATION_CLASSES"] = (
            "magnify.apps.core.authentication.DelegatedJWTAuthentication",
            "rest_framework.authentication.SessionAuthentication",
        )


class Test(Base):
    """Test environment settings"""

    JITSI_CONFIGURATION = {
        "jitsi_domain": "meeting.education",
        "jitsi_guest_avatar": "",
        "jitsi_guest_default_password": "default",
        "jitsi_guest_username": "guest",
        "jitsi_xmpp_domain": "meet.jitsi",
        "jitsi_secret_key": "ThisIsAnExampleKeyForTestPurposeOnly",
        "jitsi_app_id": "app_id",
        "jitsi_token_expiration_seconds": 300,
    }
    SIMPLE_JWT = {
        "USER_ID_FIELD": "jwt_sub",
        "USER_ID_CLAIM": "sub",
    }


class ContinuousIntegration(Test):
    """
    Continuous Integration environment settings

    nota bene: it should inherit from the Test environment.
    """

    ALLOWED_HOSTS = ["*"]
    CORS_ALLOW_ALL_ORIGINS = True
    CSRF_TRUSTED_ORIGINS = ["http://localhost:8070"]


class Production(Base):
    """Production environment settings

    You must define the DJANGO_ALLOWED_HOSTS and DJANGO_SECRET_KEY environment
    variables in Production configuration (and derived configurations):

    DJANGO_ALLOWED_HOSTS="foo.com,foo.fr"
    DJANGO_SECRET_KEY="your-secret-key"
    """

    # Security
    SECRET_KEY = values.SecretValue()
    ALLOWED_HOSTS = values.ListValue([])
    CSRF_COOKIE_SECURE = True
    SECURE_BROWSER_XSS_FILTER = True
    SECURE_CONTENT_TYPE_NOSNIFF = True
    SESSION_COOKIE_SECURE = True

    # For static files in production, we want to use a backend that includes a hash in
    # the filename, that is calculated from the file content, so that browsers always
    # get the updated version of each file.
    STORAGES = {
        "staticfiles": {
            "BACKEND": str(
                values.Value("whitenoise.storage.CompressedManifestStaticFilesStorage")
            ),
        },
    }


class Feature(Production):
    """
    Feature environment settings

    nota bene: it should inherit from the Production environment.
    """


class Staging(Production):
    """
    Staging environment settings

    nota bene: it should inherit from the Production environment.
    """


class PreProduction(Production):
    """
    Pre-production environment settings

    nota bene: it should inherit from the Production environment.
    """
